pkg/controller/deployment_util.go

```diff
@@ -618,15 +619,30 @@ func SetReplicasAnnotations(is *v1alpha1.MachineSet, desiredReplicas, maxReplica
 
 // MaxUnavailable returns the maximum unavailable machines a rolling deployment can take.
 func MaxUnavailable(deployment v1alpha1.MachineDeployment) int32 {
-	if !IsRollingUpdate(&deployment) || (deployment.Spec.Replicas) == 0 {
+	if !IsRollingUpdate(&deployment) && !IsInPlaceUpdate(&deployment) || (deployment.Spec.Replicas) == 0 {
```

Comment:
Check this condition properly. Condition `!IsRollingUpdate(&deployment)` will be true for `inplace update` while `!IsInPlaceUpdate(&deployment)` will be true for `rolling update`. Both the condition cannot occur at once.